ddtrace/llmobs/_evaluators/ragas/context_precision.py

```diff
@@ -0,0 +1,159 @@
+                                      submitting evaluation metrics.
+
+        Raises: NotImplementedError if the ragas library is not found or if ragas version is not supported.
+        """
+        super().__init__(llmobs_service)
+        self.ragas_context_precision_instance = self._get_context_precision_instance()
+        self.context_precision_output_parser = self.mini_ragas.RagasoutputParser(
+            pydantic_object=self.mini_ragas.ContextPrecisionVerification
+        )
+
+    def _get_context_precision_instance(self):
+        """
+        This helper function ensures the context precision instance used in
+        ragas evaluator is updated with the latest ragas context precision instance
+        instance AND has an non-null llm
+        """
+        if self.mini_ragas.context_precision is None:
+            return None
+        ragas_context_precision_instance = self.mini_ragas.context_precision
+        if not ragas_context_precision_instance.llm:
+            ragas_context_precision_instance.llm = self.mini_ragas.llm_factory()
+        return ragas_context_precision_instance
+
+    def evaluate(self, span_event: dict) -> Tuple[Union[float, str], Optional[dict]]:
+        """
+        Performs a context precision evaluation on an llm span event, returning either
+            - context precision score (float) OR failure reason (str)
+            - evaluation metadata (dict)
+        If the ragas context precision instance does not have `llm` set, we set `llm` using the `llm_factory()`
+        method from ragas which currently defaults to openai's gpt-4o-turbo.
+        """
+        self.ragas_context_precision_instance = self._get_context_precision_instance()
+        if not self.ragas_context_precision_instance:
+            return "fail_context_precision_is_none", {}
+
+        evaluation_metadata = {EVALUATION_KIND_METADATA: "context_precision"}  # type: dict[str, Union[str, dict, list]]
+
+        # initialize data we annotate for tracing ragas
+        score, question, answer = (
+            math.nan,
+            None,
+            None,
+        )
+
+        with self.llmobs_service.workflow(
+            "dd-ragas.context_precision", ml_app=_get_ml_app_for_ragas_trace(span_event)
+        ) as ragas_cp_workflow:
+            try:
+                evaluation_metadata[EVALUATION_SPAN_METADATA] = self.llmobs_service.export_span(span=ragas_cp_workflow)
+
+                cp_inputs = self._extract_evaluation_inputs_from_span(span_event)
+                if cp_inputs is None:
+                    logger.debug(
+                        "Failed to extract evaluation inputs from "
+                        "span sampled for `ragas_context_precision` evaluation"
+                    )
+                    return "fail_extract_context_precision_inputs", evaluation_metadata
+
+                question = cp_inputs["question"]
+                contexts = cp_inputs["contexts"]
+                answer = cp_inputs["answer"]
```

Comment:
are these separate variable declarations necessary since these are only used once? I'm fine just calling them directly from the `cp_inputs[key_name]` since it's still clear what's happening here without it

Reply:
True!